Create  ZIP402 (#373)
zcash/zips: zip-0402.rst

::

  ZIP: 402
  Title: Change the fundamental wallet database from BerkeleyDB to SQLite
  Owners: Alfredo Garcia <oxarbitrage@gmail.com>
  Status: Draft
  Category: Wallet
  Created: 2020-05-30
  License: MIT

Terminology
===========

The key words "MUST" and "MAY" in this document are to be interpreted as described in RFC 2119. [#RFC2119]_

Abstract
========

This document proposes a change of the database zcashd wallet currently use from BerkeleyDB into SQLite. It also proposes the separation of wallet transaction and other recoverable data into a separated database file.

Motivation
==========

> The main reason of the database change is reliability, SQLite haves a very good atomicity implementation [#SQLiteAtomic]_. Relational databases reduces redundancy and SQLite specifically is the world most used for local storage, it is also well mantained, have a good C++ interface, it is portable and open source.

Reviewer: ```suggestion
The main reason of the database change is reliability; SQLite has a very good atomicity implementation [#SQLiteAtomic]_. Relational databases reduces redundancy and SQLite specifically is very commonly used for local storage. It is also well mantained, has a good C++ interface, and is portable and open source.
```

> The database separation between recoverable versus non-recoverable data is to avoid writting too often to the file where the important non-recoverable data like public and private keys are located. Isolating transactions and others from the main wallet database will increase performance and reduce the risk of corruption.

Reviewer: ```suggestion
An advantage of the database separation between recoverable versus non-recoverable data, is to avoid writing too often to the file where the important non-recoverable data like public and private keys are located. Isolating transactions and other data that can be regenerated from the block chain from the keys in the main wallet database, will increase performance and reduce the risk of corruption.
```

Specification
=============

After [#ZIP400]_ and its detailed information about all database keys-values and functionality we know what is needed to change the fundamental database to something else.

> This document proposes a new relational schema for 2 new SQLite databases where all the original properties are present but just organized in a different way. Below schema is an aproximation and can change during implementation.

Reviewer 2: ```suggestion
This document proposes a new relational schema for two new SQLite databases. All the original properties currently stored using BerkleyDB are present, but are now organized in a different way. The schema below is an approximation and can change during implementation.
```

The following sql schema is proposed for a database named **wallet.sqlite** with non-recoverable information::

    CREATE TABLE accounts(
        id                INTEGER PRIMARY KEY AUTOINCREMENT
        id_address        INTEGER
        name              TEXT
        account           BLOB);
        
    CREATE TABLE account_entries(
        id                INTEGER PRIMARY KEY AUTOINCREMENT,
        id_account        INTEGER,
        entry             BLOB);

    CREATE TABLE addresses(
        id                INTEGER PRIMARY KEY AUTOINCREMENT,
        address           TEXT,
        id_account        INTEGER,
        name              TEXT,
        purpose           TEXT);
        
    CREATE TABLE destdata(
        id                INTEGER PRIMARY KEY AUTOINCREMENT,
        id_address        INTEGER,
        key               TEXT,

>         value             TEXT);

Reviewer 2: This looks vestigial; there are no calls to CWallet::AddDestData.
